src/engine/strat_engine/engine.rs

```diff
@@ -125,6 +137,73 @@ impl Engine for StratEngine {
         Ok(uuid)
     }
 
+    /// Evaluate a device node & devicemapper::Device to see if it's a valid
+    /// stratis device.  If all the devices are present in the pool and the pool isn't already
+    /// up and running, it will get setup and the pool uuid will be returned.
+    ///
+    /// Returns an error if the status of the block device can not be evaluated.
+    fn block_evaluate(&mut self,
+                      dev_node: PathBuf,
+                      device: Device)
+                      -> EngineResult<Option<PoolUuid>> {
+        let dm = DM::new()?;
+        if let Some(pool_uuid) = is_stratis_device(&dev_node)? {
+            if self.pools.contains_uuid(pool_uuid) {
+                // TODO: Handle the case where we have found a device for an already active pool.
+                warn!("Pool {} is already known, ignoring!", pool_uuid);
+                Ok(None)
+            } else {
+
+                let mut devices = self.incomplete_pools
+                    .remove(&pool_uuid)
+                    .or_else(|| Some(HashMap::new()))
+                    .expect("We just retrieved or created a HashMap");
+                devices.insert(device, dev_node);
+
+                let rc = if let Ok(pool) = StratPool::setup(&dm, pool_uuid, &devices) {
+                    // We know we have a unique uuid, but the pools table requires a unique name too
+                    // so we will ensure unique before we insert as we don't want to change the
+                    // existing state if we have a conflict.
+                    //
+                    // We will also _not_ exit the daemon as we do in initialize as we were
+                    // previously up and running for some duration of time.
+                    if !self.pools.contains_name(pool.name()) {
+                        self.pools.insert(pool);
+                        Some(pool_uuid)
+                    } else {
+                        let mut dev_paths = String::new();
```

Comment:
Simplify to something like this, please.

```
let dev_paths = devices.values().map(|p| p.to_str().expect(...)).join(", ");
```

Reply:
Done